### PR TITLE
[2731] Test setup

### PR DESCRIPTION
**Pull request by TobyRet**
### Context

We recently undertook a spike on our test suite to identify opportunities for improve, particularly around decreasing the time it takes to run the tests.

This PR acts on one the recommendations and aims to create a standard test setup process. What this means is that each test will have access to standardised set of test data (with the ability to customise where required). This will remove a lot of the duplicated factories that are instantiated across all the tests and increase overall speed. 

This work is based on similar work undertaken for [Ministry Of Justice Correspondance Tool](https://github.com/ministryofjustice/correspondence_tool_staff/blob/master/spec/support/standard_setup.rb)

### Guidance to review

NB: This PR should be considered a 'first iteration'. Functionality has been added to create a limited set of test data. The TestSetup class has been wired up to two specs. Moving forwards we would need to decide how to migrate all specs over to the TestSetup process (possibly we could add a new PR checklist requirement and migrate as we go?)

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: spec/support/test_setup.rb [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        super
+      end
+    end
+
+    @@recruitment_cycles = {
+      current_cycle: -> { FactoryBot.find_or_create :recruitment_cycle },
+      next_cycle: -> { FactoryBot.find_or_create :recruitment_cycle, :next }
+    }
+
+    @@users = {
```
Comment (TobyRet) [inline spec/support/test_setup.rb]: Would it make sense to have 'creator services' for users, orgs and providers? (similar to the `SubjectCreatorService`). This would cut down on a lot of factory creation

File: spec/models/recruitment_cycle_spec.rb [hunk 2]
```diff
@@ -13,6 +13,10 @@
 require "rails_helper"
 
 describe RecruitmentCycle, type: :model do
+
+  let(:current_cycle) { TestSetup.current_cycle }
+  let(:next_cycle) { TestSetup.next_cycle }
+
   subject { find_or_create :recruitment_cycle, year: "2019" }
```
Comment (misaka) [inline spec/models/recruitment_cycle_spec.rb]: Would it make sense, as part of this PR, to change this too? We would eventually want to do that, and since we're changing all the other cycles in this file already, maybe it would (make sense).

File: spec/requests/api/v2/providers/courses/publishable_spec.rb [hunk 3]
```diff
@@ -1,31 +1,27 @@
...
-  let(:token)        { build_jwt :apiv2, payload: payload }
-  let(:credentials) do
-    ActionController::HttpAuthentication::Token.encode_credentials(token)
+  let(:provider) { create :provider, organisations: [organisation] }
+  let(:payload) { { email: user.email } }
+  let(:token) { build_jwt :apiv2, payload: payload }
+  let(:credentials) { ActionController::HttpAuthentication::Token.encode_credentials(token) }
+  let(:setup) do
+    TestSetup.new(with_courses: {
+      course: { provider: provider },
```
Comment (misaka) [inline spec/requests/api/v2/providers/courses/publishable_spec.rb]: I think the goal should be to avoid having to do this. I think we should change the factories to work with `find_or_create :provider`, which would mean we don't have to provide a provider when creating courses, etc.

File: spec/factories/courses.rb [hunk 4]
```diff
@@ -208,14 +208,16 @@
...
+    trait :unpublished_with_primary_maths do
       transient do
-        identifier { "published_with_primary_mathematics" }
+        identifier { "unpublished_with_primary_mathematics" }
       end
 
       name { "#{identifier} course" }
 
       provider { find_or_create :provider, :published_scitt }
+      site_statuses { [build(:site_status, :new)] }
```
Comment (misaka) [inline spec/factories/courses.rb]: This will need to be associated with the correct provider. There's a few ways to do this:

1. add functionality to the `site_status` or `site` factory so that if the provider wasn't specified, to introspect the course and use it's provider, perhaps as a separate trait like `:auto_provider`
2. add the course's provider to the attributes here
3. add a trait to the `provider` factory to be able to add a site to it

... ya, this will warrant a little bit more thinking.

Comment (TobyRet) [reply spec/factories/courses.rb]: Change made here - https://github.com/DFE-Digital/manage-courses-backend/pull/1097/commits/65c368d95828c9c3b04fe6d3ffbf13d46700ec03

File: spec/support/reference_data.rb [hunk 5]
```diff
@@ -25,6 +25,14 @@
   config.before(:all) do
     SubjectCreatorService.new.execute
     SubjectFinancialIncentiveCreatorService.new.execute
```
Comment (TobyRet) [inline spec/support/reference_data.rb]: I guess the SubjectFinancialIncentiveCreatorService and SubjectCreatorService be migrated over to the TestSetup pattern? And this is where sub-classing the TestSetup class might be useful ...

File: spec/requests/api/v2/providers/courses/publishable_spec.rb [hunk 6]
```diff
@@ -69,8 +49,9 @@
     describe "failed validation" do
       let(:json_data) { JSON.parse(subject.body)["errors"] }
 
-      context "no enrichments" do
-        let(:course) { create(:course, provider: provider) }
+      context "no enrichments and location" do
+        let(:course) { TestSetup.unpublished_course_with_no_location_or_enrichment }
```
Comment (TobyRet) [inline spec/requests/api/v2/providers/courses/publishable_spec.rb]: How 'granular' should we be with regards to course test data?

File: spec/factories/courses.rb [hunk 7]
```diff
@@ -127,6 +127,11 @@
       # We've just created a course with this provider's code, so ensure it's
       # up-to-date and has this course loaded.
       course.provider.reload
+
```
Comment (TobyRet) [inline spec/factories/courses.rb]: I need to add 'find_or_create' to the provider on line 45

**Comment by TobyRet** [comment]
[Create standard test data setup](https://trello.com/c/AZMRB3wq/2731-create-standard-test-data-setup)

File: spec/requests/api/v2/providers/courses/publishable_spec.rb [hunk 8]
```diff
@@ -43,23 +32,14 @@
     include_examples "Unauthenticated, unauthorised, or not accepted T&Cs"
 
     context "when course and provider is not related" do
-      let(:course) { create(:course) }
+      let(:course)   { create(:course, :unpublished_with_primary_maths) }
+      let(:provider) { create(:provider) }
```
Comment (faissalMT) [inline spec/requests/api/v2/providers/courses/publishable_spec.rb]: Shouldn't these now be using `TestSetup`

File: spec/requests/api/v2/providers/courses/publishable_spec.rb (hunk 6 above)
Comment (faissalMT) [reply spec/requests/api/v2/providers/courses/publishable_spec.rb]: I think we should increase granularity as needed rather than upfront so as to try and have a more uniform set of test courses, but I do think some tests will require high degrees of granularity although I can't think of any off the top of my head.

File: .rubocop.yml [hunk 9]
```diff
@@ -13,3 +13,7 @@ AllCops:
 
 Rails/HasManyOrHasOneDependent:
   Enabled: false
+
+Style/ClassVars:
+  Exclude:
+    - 'spec/support/test_setup.rb'
```
Comment (misaka) [inline .rubocop.yml]: Is this your preferred way to exclude a specific test for just one file? I'm more used to including it in the file so it's more obvious.

File: spec/factories/courses.rb [hunk 10]
```diff
@@ -125,6 +125,11 @@
...
+      # Make sure this Course and and Sites are associated with the same Provider
+      if course.site_statuses.any?
+        course.site_statuses.first.site.provider = course.provider
```
Comment (misaka) [inline spec/factories/courses.rb]: Hrm. A side-effect of doing it this way, I think, is that if a test tries to associate a `site` with a `course` that is from another `provider`, say to test a validation or something similar, then this is going to over-write that.

The original comment I left was about making sure that the site created by default had the correct association. I think we should be a bit more cautious about how we do that.

File: spec/factories/courses.rb [hunk 11]
```diff
@@ -209,5 +214,17 @@
     trait :skip_validate do
       to_create { |instance| instance.save(validate: false) }
     end
+
+    trait :unpublished_with_primary_maths do
+      transient do
+        identifier { "unpublished_with_primary_mathematics" }
+      end
+
+      name { "#{identifier} course" }
```
Comment (misaka) [inline spec/factories/courses.rb]: I realised after I did this that it should actually be named after the attribute, so we can track where it gets used it if shows up elsewhere.

```suggestion
      name { "#{identifier} course name" }
```

File: spec/factories/organisations.rb [hunk 12]
```diff
@@ -15,5 +15,9 @@
   factory :organisation do
     name { "LONDON SCITT" + rand(1000000).to_s }
     sequence(:org_id)
+
+    trait :with_user do
+      users { [find_or_create(:user)] }
```
Comment (misaka) [inline spec/factories/organisations.rb]: Not so sure about the `find_or_create` here (sorry if I suggested it should be this).

Reason being I think it's more common that, by default, each organisation gets one user. If we want one user to be shared across organisations, it's my feeling we should have to do that explicitly.

File: spec/factories/providers.rb [hunk 13]
```diff
@@ -90,5 +90,13 @@
     trait :previous_recruitment_cycle do
       recruitment_cycle { find_or_create :recruitment_cycle, :previous }
     end
+
+    trait :published_scitt do
+      transient do
+        identifier { "published_scitt" }
+      end
+
+      provider_name { "#{identifier} provider" }
```
Comment (misaka) [inline spec/factories/providers.rb]:
```suggestion
      provider_name { "#{identifier} provider name" }
```

File: spec/requests/api/v2/providers/courses/publishable_spec.rb (hunk 6 above)
Comment (misaka) [reply spec/requests/api/v2/providers/courses/publishable_spec.rb]: I think we should try to list out what courses we could reasonably expect to see in the wild. For stuff like failure case testing, maybe a single course that doesn't have any values set might be enough ... trying to cater to each test's requirements with a TestSetup course would be overkill.

File: spec/support/reference_data.rb (hunk 5 above)
Comment (misaka) [reply spec/support/reference_data.rb]: Well, the thing is that `SubjectCreatorService` and `SubjectFinancialIncentiveCreatorService` are used by the db seeding. I wouldn't want db seeding to use factories, so moving them here would imply duplicating whatever logic is in those services between seeding and factories.

File: spec/support/reference_data.rb [hunk 14]
```diff
@@ -25,6 +25,14 @@
...
+
+    # Creating or finding course test data
+    if Settings.test_setup
+      TestSetup.unpublished_course_with_primary_maths
+      TestSetup.unpublished_course_with_no_location_or_enrichment
+      TestSetup.unpublished_course_with_draft_enrichment
+      TestSetup.unpublished_fee_type_based_course_with_invalid_enrichment
+    end
```
Comment (misaka) [inline spec/support/reference_data.rb]: Isn't this done already by instantiating `TestSetup`?

File: spec/support/test_setup.rb [hunk 15]
```diff
@@ -0,0 +1,90 @@
+class TestSetup
+  # Class that instantiates standard set of test data for specs.
+
+  @@recruitment_cycles = {
+    current_cycle: -> { FactoryBot.find_or_create :recruitment_cycle },
```
Comment (misaka) [inline spec/support/test_setup.rb]: Can we get by without using `FactoryBot` here? It seems really noisy and affects reability, imo.

File: .rubocop.yml (hunk 9 above)
Comment (TobyRet) [reply .rubocop.yml]: No problem - I will inline the exclusion

File: spec/factories/organisations.rb (hunk 12 above)
Comment (TobyRet) [reply spec/factories/organisations.rb]: Ok makes sense

File: spec/support/reference_data.rb [hunk 16]
```diff
@@ -25,6 +25,14 @@
   config.before(:all) do
     SubjectCreatorService.new.execute
     SubjectFinancialIncentiveCreatorService.new.execute
+
+    # Creating or finding course test data
+    if Settings.test_setup
```
Comment (misaka) [inline spec/support/reference_data.rb]: Good call using `Settings`

**Comment by misaka** [comment]
I think we still need to work on not creating too many extra records here ... it looks like for every `unpublished_course_with_primary_maths` being created, 4 organisations, 4 users and 2 providers are being created ... and that's for each time the course is being created too.

I tested this by adding a quick example to do some manual testing to `spec/requests/api/v2/providers/courses/publishable_spec.rb`:

```
    it "tests" do
      binding.pry
      nil
    end
```

And then looking at the number of records being created:

```ruby
[1] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> Organisation.count
=> 0
[2] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> Provider.count
=> 0
[3] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> User.count
=> 0
[4] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> Course.count
=> 0
[5] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> TestSetup.unpublished_course_with_primary_maths
=> #<Course:0x00007fe36147b9a8
 id: 1,
 ...>
[6] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> Organisation.count
=> 4
[7] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> Provider.count
=> 2
[8] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> User.count
=> 4
[9] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> Course.count
=> 1
[10] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> TestSetup.unpublished_course_with_primary_maths
=> #<Course:0x00007fe3614d26e0
 id: 2,
 ...>
[11] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> Organisation.count
=> 8
[12] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> Provider.count
=> 4
[13] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> User.count
=> 8
[14] pry(#<RSpec::ExampleGroups::PublishableAPIV2::POSTPublishable>)> Course.count
=> 2
```

I would expect to only see 1 organisation, 1 user and 1 provider (I don't think we should be creating a course with an accredited body by default) being created, and probably, I think, they should be shared across the courses so that we don't create extra resources by default (we explicitly have to create extra providers if we want courses from different providers to test with).

File: spec/requests/api/v2/providers/courses/publishable_spec.rb [hunk 17]
```diff
@@ -43,23 +32,14 @@
...
-      let!(:course) do
-        create(:course,
-               provider: provider,
-               site_statuses: [site_status],
-               enrichments: [enrichment],
-               age: 17.days.ago,
-               subjects: [primary_with_mathematics])
-      end
+    context "unpublished course with draft enrichment" do
+      let(:course) { TestSetup.unpublished_course_with_draft_enrichment }
```
Comment (timabell) [inline spec/requests/api/v2/providers/courses/publishable_spec.rb]: Is the intention that we stop using `build`/`create` entirely? Or will still use build/create for interesting edge cases

File: spec/requests/api/v2/providers/courses/publishable_spec.rb (hunk 6 above)
Comment (timabell) [reply spec/requests/api/v2/providers/courses/publishable_spec.rb]: I'm also a bit concerned about the proliferation of methods like this for every variation. I think we should set a strong opinion out as to what in what cases `TestSetup` should be extended vs. just building custom things. `unpublished_course_with_no_location_or_enrichment` smells to me a bit of heading down the proliferation path already and maybe we shouldn't change this one?

File: spec/requests/api/v2/providers/courses/publishable_spec.rb [hunk 18]
```diff
@@ -81,15 +62,8 @@
...
-        let(:course) {
-          create(:course, :fee_type_based,
-                 provider: provider,
-                 enrichments: [invalid_enrichment],
-                 site_statuses: [site_status])
-        }
-
         context "invalid enrichment with invalid content lack_presence fields" do
-          let(:invalid_enrichment) { create(:course_enrichment, :without_content) }
+          let(:course) { TestSetup.unpublished_fee_type_based_course_with_invalid_enrichment }
```
Comment (timabell) [inline spec/requests/api/v2/providers/courses/publishable_spec.rb]: As above, smells of unbounded proliferation coming down the line

File: spec/requests/api/v2/providers/courses/publishable_spec.rb (hunk 17 above)
Comment (misaka) [reply spec/requests/api/v2/providers/courses/publishable_spec.rb]: I think that that's a re-stating of the granularity question above, and ya ... we would still use create/build for interesting edge cases, aiming to just cover the majority of the types of courses in `TestSetup` but not all.

File: spec/support/test_setup.rb [hunk 19]
```diff
@@ -0,0 +1,90 @@
+class TestSetup
+  # Class that instantiates standard set of test data for specs.
```
Comment (timabell) [inline spec/support/test_setup.rb]: Would be good to mention what problem we are solving by having this

File: spec/support/test_setup.rb [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+      }.merge(attributes))
+    end,
+  }
+
+  # Used when not instantiating a TestSetup object in a before block.
+  # This will create a new fixture, using 'find_or_create' for teams and
+  # users and 'create' for cases. This is used to have a standardised set of
+  # recruitment cycles to work with, but without requiring
+  # pre-instantiation.
+
```
Comment (timabell) [inline spec/support/test_setup.rb]: Is the above the docs for `def self.method_missing`? If so kill the extra newline

File: spec/support/test_setup.rb [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+        enrichments: [FactoryBot.build(:course_enrichment, :without_content)],
+      }.merge(attributes))
+    end,
+  }
+
+  # Used when not instantiating a TestSetup object in a before block.
+  # This will create a new fixture, using 'find_or_create' for teams and
+  # users and 'create' for cases. This is used to have a standardised set of
+  # recruitment cycles to work with, but without requiring
+  # pre-instantiation.
```
Comment (timabell) [inline spec/support/test_setup.rb]: I'm not sure I entirely understand this

File: spec/support/test_setup.rb [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+    if @@recruitment_cycles&.key?(method_name)
+      @@recruitment_cycles[method_name].call
+    elsif @@courses&.key?(method_name)
+      @@courses[method_name].call
+    else
+      super
+    end
+  end
+
+  def self.respond_to_missing?(method_name, include_private = false)
```
Comment (timabell) [inline spec/support/test_setup.rb]: I'm assuming the comment above the above method applies to this too. Might be worth documenting them individually too as they're pretty hard to parse for someone like me.

File: spec/support/test_setup.rb [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+    @@recruitment_cycles&.key?(method_name) ||
+      @@courses&.key?(method_name) ||
+      super
+  end
+
+  attr_reader :courses, :recruitment_cycles
+
+  def initialize(with_courses: nil)
```
Comment (timabell) [inline spec/support/test_setup.rb]: Shouldn't this be at the top of the class?

File: spec/support/test_setup.rb [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+    else
+      course_types = with_courses || @@courses.keys
+      @courses = Hash[
+        @@courses.slice(*course_types).map do |name, kourse|
+          # instantiate course by calling the blocks in @@courses
+          [name, kourse.call]
+        end
+      ]
+    end
+  end
```
Comment (timabell) [inline spec/support/test_setup.rb]: This if/else block is pretty confusing to me. Perhaps a comment, or refactor out into named functions?

File: spec/support/test_setup.rb [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+
+  def method_missing(method_name, *args)
```
Comment (timabell) [inline spec/support/test_setup.rb]: As above; Might be worth documenting `method_missing` and `respond_to_missing` too as I find them both pretty hard to parse.

File: spec/factories/providers.rb [hunk 26]
```diff
@@ -90,5 +90,13 @@
     trait :previous_recruitment_cycle do
       recruitment_cycle { find_or_create :recruitment_cycle, :previous }
     end
+
+    trait :published_scitt do
+      transient do
+        identifier { "published_scitt" }
```
Comment (defong) [inline spec/factories/providers.rb]: is this a glorified variable?

File: spec/support/test_setup.rb [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+    end,
+    unpublished_fee_type_based_course_with_invalid_enrichment: ->(attributes = {}) do
+      find_or_create(:course, :fee_type_based, :unpublished_with_primary_maths, {
+        enrichments: [build(:course_enrichment, :without_content)],
+      }.merge(attributes))
+    end,
+  }
+
+  # Used when not instantiating a TestSetup object in a before block.
+  # This will create a new fixture, using 'find_or_create' for teams and
```
Comment (defong) [inline spec/support/test_setup.rb]: lost... what is teams refering to?

File: spec/factories/providers.rb (hunk 26 above)
Comment (TobyRet) [reply spec/factories/providers.rb]: Indeed it is

File: spec/factory_specs/course_spec.rb [hunk 28]
```diff
@@ -30,4 +30,23 @@
...
+  context "unpublished_with_primary_maths" do
+  subject! { TestSetup.unpublished_course_with_primary_maths }
+
+    its(:name) { should eq("unpublished_with_primary_maths course name")}
+
+    it "has the correct number of associations" do
+      expect(Site.count).to eq(1)
+      expect(Provider.count).to eq(1)
+      expect(Organisation.count).to eq(1)
+      expect(User.count).to eq(1)
```
Comment (misaka) [inline spec/factory_specs/course_spec.rb]: When we add `TestSetup` to the `before :all`, this fail. Probably better to change `subject!` to `subject` and do:

```
      expect { subject }.not_to change { Site.count }
```

... etc

File: spec/factory_specs/course_spec.rb [hunk 29]
```diff
@@ -30,4 +30,23 @@
       expect(CourseEnrichment.where("created_at < ?", 4.days.ago).count).to eq(1)
     end
   end
+
+  context "unpublished_with_primary_maths" do
+  subject! { TestSetup.unpublished_course_with_primary_maths }
```
Comment (misaka) [inline spec/factory_specs/course_spec.rb]:
```suggestion
    subject! { TestSetup.unpublished_course_with_primary_maths }
```

File: spec/factory_specs/course_spec.rb (hunk 28 above)
Comment (misaka) [reply spec/factory_specs/course_spec.rb]: Also, I think the factory we're testing here should only create one `Site` and `Provider`, the `provider` factory should only create one `Organisation`, and the `organisation` factory should only create one `User`.

File: spec/models/provider_spec.rb [hunk 30]
```diff
@@ -300,7 +300,7 @@
     describe ".include_courses_counts" do
       let(:first_provider) { Provider.include_courses_counts.first }
       before do
-        provider
+        provider.courses << course
```
Comment (misaka) [inline spec/models/provider_spec.rb]: This should be handled in the factory.

File: spec/requests/api/v2/providers/courses/publishable_spec.rb (hunk 6 above)
Comment (misaka) [reply spec/requests/api/v2/providers/courses/publishable_spec.rb]: Ya, I agree that this one sounds a bit too specialised.

File: spec/services/providers/copy_to_recruitment_cycle_service_spec.rb [hunk 31]
```diff
@@ -37,6 +36,10 @@
       )
     end
 
+    before do
+      provider.courses << course
+    end
```
Comment (misaka) [inline spec/services/providers/copy_to_recruitment_cycle_service_spec.rb]: As above somewhere, I think we need to get this into the factory.

File: spec/support/test_setup.rb (hunk 23 above)
Comment (misaka) [reply spec/support/test_setup.rb]: This feels alright to me. Everything above are class-scoped so this is where the instance-specific stuff starts.

File: spec/support/test_setup.rb [hunk 32]
```diff
@@ -0,0 +1,94 @@
+# rubocop:disable Style/ClassVars
+class TestSetup
```
Comment (misaka) [inline spec/support/test_setup.rb]: iirc, you can just do this:

```
class TestSetup # rubocop:disable Style/ClassVars
```

and it'll only disable those checks for this class.

**Comment by misaka** [comment]
@TobyRet did his presentation of this to the team and we agreed we'd continue to look at this approach. First step is to try to improve the API to not use method missing, but to use something like:

```
@setup = TestSetup.new
@setup.course :unpublished, :primary, :mathematics
```
